### PR TITLE
[Loss] Fix loss layer were allocating a new mem @open sesame 02/19 18:25

### PR DESCRIPTION
**Pull request by zhoonit**
- [Tensor/Clean] Relocate tensor methods
- [Tensor] rearrange methods
- [Loss] Fix loss layer were allocating a new mem
```
As allocation is managed in `manager`, layer shouldn't be allocate a new
memory that are managed by memory.
However, loss layer was allocating a new memory. This patch fixes the issue.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>
Cc: Parichay Kapoor <pk.kapoor@samsung.com>
```

### CODE REVIEW

File: nntrainer/layers/loss_layer.cpp [hunk 1]
```diff
@@ -128,15 +128,31 @@ void LossLayer::calcDerivative() {
 
   switch (loss_type) {
   case LossType::LOSS_MSE:
-    ret_derivative = y.subtract(y2).multiply(2).divide(y.getDim().getDataLen());
+    y.subtract(y2, ret_derivative);
+    ret_derivative.multiply_i(2);
+    if (ret_derivative.divide_i(y.length()) != ML_ERROR_NONE) {
```
Comment (zhoonit) [inline nntrainer/layers/loss_layer.cpp:133]: Please do not review yet, this patch is meant for the testing purpose as of now, until the underlying bug is fixed.

**Comment by kparichay** [comment]
@zhoonit The reason for the failure of the PR has been identified and it is due to yet another bug in in-place layer optimization.
This will be fixed in a couple of days as it requires some major changes.

**Comment by kparichay** [comment]
@zhoonit Now that #880 has been merged, can you please rebase this and check if any issue still exists?

**Comment by zhoonit** [comment]
> @zhoonit Now that #880 has been merged, can you please rebase this and check if any issue still exists?

Done.

File: nntrainer/layers/loss_layer.cpp (hunk 1 above)
Comment (kparichay) [reply nntrainer/layers/loss_layer.cpp:134]: I guess this is ready to review now?

File: nntrainer/layers/loss_layer.cpp [hunk 2]
```diff
@@ -129,15 +129,31 @@ void LossLayer::calcDerivative() {
...
+      throw std::runtime_error(
+        "[Loss::calcDerivative] Error when calculating loss");
+    }
+    // ret_derivative = ret_derivative.divide(y.length());
     break;
   case LossType::LOSS_ENTROPY_SIGMOID:
-    ret = y.apply(ActivationLayer::sigmoid, ret);
-    ret_derivative = ret.subtract(y2).divide(ret.getDim().getDataLen());
+    y.apply(ActivationLayer::sigmoid, ret);
+    ret.subtract(y2, ret_derivative);
```
Comment (kparichay) [inline nntrainer/layers/loss_layer.cpp:142]:
```suggestion
    y.apply(ActivationLayer::sigmoid, ret_derivative);
    ret_derivative.subtract_i(y2);
```

File: nntrainer/layers/loss_layer.cpp [hunk 3]
```diff
@@ -129,15 +129,31 @@ void LossLayer::calcDerivative() {
...
+    if (ret_derivative.divide_i(ret.length()) != ML_ERROR_NONE) {
+      throw std::runtime_error(
+        "[Loss::calcDerivative] Error when calculating loss");
+    }
+    // ret_derivative = ret.subtract(y2).divide(ret.getDim().getDataLen());
     break;
   case LossType::LOSS_ENTROPY_SOFTMAX:
-    ret = y.apply(ActivationLayer::softmax, ret);
-    ret_derivative = ret.subtract(y2).divide(ret.batch());
+    y.apply(ActivationLayer::softmax, ret);
```
Comment (kparichay) [inline nntrainer/layers/loss_layer.cpp:150]: Please add a comment here that y and ret_derivative can be same here. 
So, this operation cannot apply in-place.

File: nntrainer/layers/loss_layer.cpp [hunk 4]
```diff
@@ -129,15 +129,31 @@ void LossLayer::calcDerivative() {
...
+    ret.subtract(y2, ret_derivative);
+    if (ret_derivative.divide_i(ret.batch()) != ML_ERROR_NONE) {
+      throw std::runtime_error(
+        "[Loss::calcDerivative] Error when calculating loss");
+    }
+    // ret_derivative = ret.subtract(y2).divide(ret.batch());
```
Comment (kparichay) [inline nntrainer/layers/loss_layer.cpp:156]: delete?

File: nntrainer/layers/loss_layer.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/layers/loss_layer.cpp:150]: https://github.com/nnstreamer/nntrainer/issues/943#issuecomment-773865415

File: nntrainer/layers/loss_layer.cpp (hunk 1 above)
Comment (zhoonit) [reply nntrainer/layers/loss_layer.cpp:134]: Yup

File: nntrainer/layers/loss_layer.cpp (hunk 3 above)
Comment (zhoonit) [reply nntrainer/layers/loss_layer.cpp:150]: Done

File: nntrainer/layers/loss_layer.cpp (hunk 4 above)
Comment (zhoonit) [reply nntrainer/layers/loss_layer.cpp:156]: Done

File: nntrainer/layers/loss_layer.cpp (hunk 2 above)
Comment (zhoonit) [reply nntrainer/layers/loss_layer.cpp:142]: Done

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Review by kparichay — APPROVED** [review]
LGTM

**Pull request merged**